euphonic/spectra.py

```diff
@@ -669,10 +674,387 @@ def broaden(self: T, x_width,
+
+        e.g. for Spectrum2DCollection this is
+
+            {"x_data": self.x_data, "y_data": self.y_data}
+        """
+        return {f"{axis}_data": getattr(self, f"{axis}_data")
+                for axis in self._bin_axes}
+
+    @classmethod
+    def _get_item_data(cls, item: Spectrum) -> Quantity:
+        return getattr(item, f"{cls._spectrum_axis}_data")
+
+    @classmethod
+    def _get_item_raw_data(cls, item: Spectrum) -> np.ndarray:
+        return getattr(item, f"_{cls._spectrum_axis}_data")
+
+    @classmethod
+    def _get_item_data_unit(cls, item: Spectrum) -> str:
+        return getattr(item, f"{cls._spectrum_axis}_data_unit")
+
+    def sum(self) -> Spectrum:
+        """
+        Sum collection to a single spectrum
+
+        Returns
+        -------
+        summed_spectrum
+            A single combined spectrum from all items in collection. Any
+            metadata in 'line_data' not common across all spectra will be
+            discarded
+        """
+        metadata = copy.deepcopy(self.metadata)
+        metadata.pop('line_data', None)
+        metadata.update(self._tidy_metadata())
+        summed_s_data = ureg.Quantity(
+            np.sum(self._get_raw_spectrum_data(), axis=0),
+            units=self._get_internal_spectrum_data_unit()
+        ).to(self._get_spectrum_data_unit())
+        return self._item_type(
+            **self._get_bin_kwargs(),
+            **{self._spectrum_data_name(): summed_s_data},
+            x_tick_labels=copy.copy(self.x_tick_labels),
+            metadata=metadata
+        )
+
+    # Required methods
+    @classmethod
+    @abstractmethod
+    def from_spectra(cls, spectra: Sequence[Spectrum]) -> Self:
+        """Construct spectrum collection from a sequence of components"""
+        ...
+
+    # Mixin methods
+    def __len__(self):
+        return self._get_raw_spectrum_data().shape[0]
+
+    @overload
+    def __getitem__(self, item: int) -> Spectrum:
+        ...
+
+    @overload  # noqa: F811
+    def __getitem__(self, item: slice) -> Self:
+        ...
+
+    @overload  # noqa: F811
+    def __getitem__(self, item: Union[Sequence[int], np.ndarray]) -> Self:
+        ...
+
+    def __getitem__(
+            self, item: Union[Integral, slice, Sequence[Integral], np.ndarray]
+    ):  # noqa: F811
+        self._validate_item(item)
+        init_kwargs = {
+            self._spectrum_data_name(): self._get_spectrum_data()[item, :],
+            "x_tick_labels": self.x_tick_labels,
+            "metadata": self._get_item_metadata(item)
+                       } | self._get_bin_kwargs()
+
+        if isinstance(item, Integral):
+            return self._item_type(**init_kwargs)
+
+        return type(self)(**init_kwargs)
+
+    def _validate_item(self, item: Integral | slice | Sequence[Integral] | np.ndarray
+                       ) -> None:
+        """Raise Error if index has inappropriate typing/range"""
+        if isinstance(item, Integral):
+            return
```

Comment:
I see what you mean. The issue is that the cases need to raise different exception types and messages, so the return type would need to be something like `Tuple[bool,Exception|None,str]`. The point of breaking out this method is to avoid clutter in the main function, so if we need _too_ much logic there it might as well all be inlined.

Reply:
Perhaps it is better think of it as "raises, or doesn't" as the `None` is not used? Maybe a function name `_raise_error_if_invalid_item()` would be clearer?

Reply:
Yeah, I get what you mean. I think `validate` is fine as a name for it, like I said it was this weird feeling that there was code-smell somewhere, but I couldn't quite pick it out.

I think maybe documenting the raises in the docstring as a block might help.

Reply:
Ok, I'll fill out the docstring properly